doc/downsampling.md

```diff
@@ -0,0 +1,129 @@
+# Downsampling
+
+Downsampling is a process of reducing the number of samples stored for a time series. 
+This is necessary for the following reasons:
+
+* To control storage requirements at scale 
+* Because most of ingested data is never queried and we would like to reduce storage
+* Even if we are able to store data at highest granularity for very long periods, 
+  presenting the data on a screen with limited pixel count requires it to be downsampled 
+  when larger time range trends are to be presented.
+* Processing large amount of samples over long periods (months/year) with query engine is
+  inefficient anyway. Working with downsampled data is much faster.
+
+Note that downsampling process does not reduce the number of time series stored. It reduces
+the number of samples stored for a time series per unit time.
+
+FiloDB supports downsampling of ingested time series data at configurable resolutions. 
+Multiple downsampling algorithms can be configured. 
+
+## Configuration
+
+A dataset can be configured with one or more downsamplers. Following downsamplers are supported:
+
+* __dMin(col):__ Calculates minimum of values in column id 'col' in the downsample period.
+* __dMax(col):__ Calculates maximum of values in column id 'col' in the downsample period.
+* __dSum(col):__ Calculates sum of values in column id 'col' in the downsample period.
+* __dCount(col):__ Calculates count of values in column id 'col' in the downsample period.
+* __dAvg(col):__ Calculates average of values in column id 'col' in the downsample period.
+* __dAvgAc(avgCol,countCol):__ Calculates average using averages and counts in column ids 'avgCol' and 'countCol' respectively in the downsample period.
+* __dAvgSc(sumCol,countCol):__ Calculates average using sums and counts in column ids 'sumCol' and 'countCol' respectively in the downsample period.
+* __tTime(col):__ Chooses last timestamp of ingested sample for the downsample period 
+
+Downsamplers beginning with 'd' emit a double value in the downsampled record. Downsamplers beginning with 't'
+emit timestamp value. One could later have downsamplers that emit histogram or long columns too.
+
+Downsampling for prometheus counters will come soon.
+
+Downsampling is configured at the time of dataset creation. For example:
+
+```
+./filo-cli -Dconfig.file=conf/timeseries-filodb-server.conf  --command create --dataset prometheus --dataColumns timestamp:ts,value:double --partitionColumns tags:map --shardKeyColumns __name__,app --downsamplers "tTime(0),dMin(1),dMax(1),dSum(1),dCount(1),dAvg(1)"
+```
+
+In the above example, the data column `value` with index 1 is configured with the dMin, dMax, sSum, dCount and dAvg downsamplers.
+
+Additional configuration is supplied via the ingestion config file for the dataset.
+
+```
```

Comment:
I suggest we omit this and just have the user refer to (or supply a link to) the timeseries-dev-source file.  In the past when I tried maintaining this it always gets out of date.